### PR TITLE
Postgres fsm

### PR DESCRIPTION
**Pull request by felixge**
Initial draft. Looking forward to feedback before publishing!

Rendered version: https://deploy-preview-4--felixge.netlify.com/2017/07/23/implementing-state-machines-in-postgresql.html

### CODE REVIEW

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+FROM
+  generate_series('2017-07-23'::date, '2017-07-26', '1 day') date,
+  LATERAL (
+    SELECT order_id, order_events_fsm(event ORDER BY id) AS state
+    FROM order_events
+    WHERE time < date + '1 day'::interval
+    GROUP BY 1
+  ) orders
+GROUP BY 1, 2
+ORDER BY 1, 2
```
Comment (johannesboyne) [inline _drafts/implementing-state-machines-in-postgresql.md]: just for copy-pasting
```sql
ORDER BY 1, 2;
```
adding the semicolon here

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+ 2017-07-23 00:00:00 | awaiting_payment
+ 2017-07-24 00:00:00 | awaiting_shipment
+ 2017-07-25 00:00:00 | awaiting_refund
+ 2017-07-26 00:00:00 | canceled
+```
+
+Or perhaps even more interesting, we can use a
+[Lateral](https://www.postgresql.org/docs/9.6/static/queries-table-expressions.html#QUERIES-LATERAL)
+sub-query to break down the number of orders per state for each day of a given
+date range:
```
Comment (johannesboyne) [inline _drafts/implementing-state-machines-in-postgresql.md]: Maybe it is worth mentioning, that this is for example the _analytical power_ described before, e.g.:

> Or perhaps even more interesting, *especially for later analytics*, we can use a [...]

Furthermore one might link to the `generate_series` SRF as well, e.g.:

> we can use a [series generation function](https://www.postgresql.org/docs/9.6/static/functions-srf.html) combined with a [Lateral](https://www.postgresql.org/docs/9.6/static/queries-table-expressions.html#QUERIES-LATERAL) sub-query to [...]

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
+---
+layout: post
+title: "Implementing State Machines in PostgreSQL"
+---
+
+[Finite-state machines](https://en.wikipedia.org/wiki/Finite-state_machine)
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: I'd use singular here: "Finite-state machine _is_ a wonderful [...] that _has_" Currently it's "they [plural] are a [singular]". Yeah, possible, but I think the singular-singular version is smoother (and wikipedia uses it) :)

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+new state, and an initial state.
+
+In practice this is usually done by drawing up a directed graph that shows how
+the different states are connected to each other via events:
+
+<img src="{% asset_path orders.svg %}" alt="Finite-state machine for an order management system in PostgreSQL" />
+
+Besides guiding our implementation, these graphs can be very useful for
+discussions and analysis. For example you can immediately see how there is no
+way for a customer to return an order after it has shipped yet. Additionally
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: Remove "yet"

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+the process of naming the involved states and events automatically creates a
+precise language that can be adopted by all stakeholders.
+
+If you paid close attention so far, you might have noticed that the graph above
+does not show the combination of all states and events which simply implies
+that the missing combinations lead to an error state. The [complete FSM]({%
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: I'd change this into this:

"You might have noticed that the graph above does not show all possible combinations of all states and all events. The reason is that the missing combinations lead to error states."

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+asset_path orders-full.svg %}) can be shown as a graph as well, but I don't
+think it's very useful in practice.
+
+Anyway, as promised in the title of this article we're going to implement this
+FSM in PostgreSQL. This may not be everybody's cup of tea, but you might like
+how this approach allows us gives advanced analytical powers as a free
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: "how this approach allows us gives" -> "how this approach gives us"

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+by-product. Embedding this kind of logic into the DB can also help protecting
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: "help protecting" -> "help protect"

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+      CASE event
+        WHEN 'refund' THEN 'canceled'
+        ELSE 'error'
+      END
+    ELSE 'error'
+  END
+$$;
+```
+
+And before we proceed, let's check a few examples to make sure the function is
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: "let's check a few examples" -> "let's test with a few examples"

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+working:
+
+```sql
+SELECT state, event, order_events_transition(state, event)
+FROM (VALUES
+  ('start', 'create'),
+  ('awaiting_payment', 'pay'),
+  ('awaiting_payment', 'cancel'),
+  ('awaiting_payment', 'ship')
+) examples(state, event);
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: SQL-Peasant asks: where does `examples` come from here?

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+```
+
+The above looks correct, but it's not immediately clear how we could use this
+function to enforce our FSM on all rows for the same `order_id` in the
+`order_events` table.
+
+The first thing we need is a good way to take a list of events and call our
+transition function on them recursively to determine the resulting state. There
+are multiple ways to accomplish this in PostgreSQL, but perhaps the most
+elegant is a [user-defined aggregate](https://www.postgresql.org/docs/current/static/xaggr.html).
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: holy shit, this is sexy 🕺 💃

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+In a nutshell, a user defined aggregate is a function that has an internal
+value (state) that is updated for each input value that is being passed into it
+using a state transition function. This means it fits our FSM model like a
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: I think it's "fits _on_ our [...]"

File: _drafts/implementing-state-machines-in-postgresql.md (hunk 9 above)
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: Ahhh, figured it out! :) Minor suggestions: I'd use `AS examples(state, event)` here to maybe not trip other SQL-Peasants up

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+```
+```
+ id | order_id | event
+----+----------+--------
+  1 |        1 | create
+  2 |        1 | pay
+  3 |        1 | ship
+```
+
+If you're still on the fence about embedding this kind of logic into your db,
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: Here it's "db", earlier it was "DB". Recommendation: write "database" everywhere :)

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+let's have a look how our approach gives us advanced analytical powers as a
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: "let's have a look how" -> "let's see how"

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+
+  (3, 'create', '2017-07-23 00:00:00'),
+  (3, 'pay', '2017-07-24 00:00:00'),
+  (3, 'cancel', '2017-07-25 00:00:00'),
+  (3, 'refund', '2017-07-26 00:00:00');
+```
+
+Using our `order_events_fsm` as a [window
+function](https://www.postgresql.org/docs/9.6/static/tutorial-window.html), we
+can easily get the state history of a given order:
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: dude, nice 😮

File: _drafts/implementing-state-machines-in-postgresql.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+```
+        time         | order_events_fsm
+---------------------+-------------------
+ 2017-07-23 00:00:00 | awaiting_payment
+ 2017-07-24 00:00:00 | awaiting_shipment
+ 2017-07-25 00:00:00 | awaiting_refund
+ 2017-07-26 00:00:00 | canceled
+```
+
+Or perhaps even more interesting, we can use a
```
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: This sentence feels kinda "clunky". I'd change it to something like "Even perhaps more interesting is using a Lateral sub-query [...]"

File: _drafts/implementing-state-machines-in-postgresql.md (hunk 2 above)
Comment (mrnugget) [inline _drafts/implementing-state-machines-in-postgresql.md]: 👍  on the second point

**Comment by felixge** [comment]
@johannesboyne @mrnugget thank you so much for taking the time to have a look, and all the kind words! I'll try to integrate all your feedback this afternoon :).

File: _drafts/implementing-state-machines-in-postgresql.md (hunk 15 above)
Comment (felixge) [inline _drafts/implementing-state-machines-in-postgresql.md]: Changed it.

File: _drafts/implementing-state-machines-in-postgresql.md (hunk 1 above)
Comment (felixge) [inline _drafts/implementing-state-machines-in-postgresql.md]: Done!

**Pull request merged**